### PR TITLE
Feature 30104 deprecate 2x role fmts

### PR DESCRIPTION
**Pull request by benjaminran**
Hello,

This is a shot at #30104.

Removed the remaining support for2x role formats from file or native role definitions. Now any role definitions with "fields" params should be rejected as invalid.

Unit tests I've modified still pass, but I'm having trouble getting ./gradlew check to pass, even without my changes applied. Runs are taking >30min, I can hear my laptop start working very hard, and I'm getting errors like > Failed to start oldEs1Fixture and > node{:x-pack:plugin:watcher:qa:rest:integTest-0} failed to wait for ports files after 120000 MILLISECONDS (usually one of those, varies per run). Still looking into that - lmk if there are extreme system requirements I might not be meeting (macos 10.12, 16gb ram) or other tips. Thanks.

### CODE REVIEW

**Comment by jkakavas** [comment]
Thank you @benjaminran!  and apologies it took me so long to get to this. I will spin CI tests now and will take a look at the PR in the next couple of days

**Comment by jkakavas** [comment]
@elasticmachine test this please

**Comment by jkakavas** [comment]
@elasticmachine update branch

**Comment by jkakavas** [comment]
@elasticmachine test this please

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/permission/FieldPermissionsTests.java [hunk 1]
```diff
@@ -144,50 +144,32 @@ public void testParseFieldPermissions() throws Exception {
...
-                "\"fields\": [\"f1\", \"f2\"]" +
-                "}]}";
-        RoleDescriptor rd = RoleDescriptor.parse("test", new BytesArray(q), true,
-                XContentType.JSON);
-        assertArrayEquals(rd.getIndicesPrivileges()[0].getGrantedFields(),
-                new String[]{"f1", "f2"});
-        assertNull(rd.getIndicesPrivileges()[0].getDeniedFields());
-
-        final String failingQuery = q;
+        final String q = "{\"indices\": [ {\"names\": \"idx2\", \"privileges\": [\"p3\"], " +
```
Comment (jkakavas) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/permission/FieldPermissionsTests.java:147]: I think we can remove this method now. It doesn't check backwards compatibility anymore (as we remove BWC in this PR) and we can get the assurance that we don't parse roles in 2x format in a unit test we can add in RoleDescriptorTests.java

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java [hunk 2]
```diff
@@ -541,15 +539,19 @@ public void testUsageStats() throws Exception {
         assertThat(usageStats.get("dls"), is(flsDlsEnabled));
     }
 
-    // test that we can read a role where field permissions are stored in 2.x format (fields:...)
-    public void testBWCFieldPermissions() throws IOException {
+    // test that we reject a role where field permissions are stored in 2.x format (fields:...)
+    public void testBWCFieldPermissions() throws Exception {
```
Comment (jkakavas) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/FileRolesStoreTests.java:543]: I think we can remove this method now. It doesn't check backwards compatibility anymore (as we remove BWC in this PR) and we can get the assurance that we don't parse roles in 2x format in a unit test we can add in RoleDescriptorTests.java

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStoreTests.java [hunk 3]
```diff
@@ -78,18 +82,19 @@ public void terminateThreadPool() throws Exception {
         terminate(threadPool);
     }
 
-    // test that we can read a role where field permissions are stored in 2.x format (fields:...)
-    public void testBWCFieldPermissions() throws IOException {
+    // test that we reject a role where field permissions are stored in 2.x format (fields:...)
+    public void testBWCFieldPermissions() throws Exception {
```
Comment (jkakavas) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStoreTests.java:86]: I think we can remove this method now. It doesn't check backwards compatibility anymore (as we remove BWC in this PR) and we can get the assurance that we don't parse roles in 2x format in a unit test we can add in RoleDescriptorTests.java

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java [hunk 4]
```diff
@@ -28,8 +28,8 @@ public void testBWCFieldPermissions() throws Exception {
         try (Client client = new NoOpClient("testBWCFieldPermissions")) {
```
Comment (jkakavas) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java:28]:
```suggestion
        try (Client client = new NoOpClient("test2xFieldPermissions")) {
```

**Review by jkakavas — CHANGES_REQUESTED** [review]
Thank you for your effort @benjaminran , I left some comments and suggestions. 

We also need to think about what this would mean for people updating to 8 as they might have a cluster which was upgraded consecutively from 2.x ->5.x -> 6.x -> 7.x -> 8.x and where up to 7.x , the roles that had the `fields` field would be happily allowed and working and will now suddenly silently not apply. I'll get back to you with ideas next week once I have a few more hours to think this through

**Comment by benjaminran** [comment]
Thanks for the review, I'll try to get an update out in the next couple days

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/permission/FieldPermissionsTests.java (hunk 1 above)
Comment (benjaminran) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/permission/FieldPermissionsTests.java]: yes, makes sense. Replaced these testBWC* tests with a test case in RoleDescriptorTest.testParse

**Review by benjaminran — COMMENTED** [review]
I accidentally pushed the last couple commits with the wrong git user.email :\\. Should I force push a correction? I guess github would probably keep the first commit's user.email (the correct one in this case) when it squashes, but don't know if your cla-checker would let things get that far.

> We also need to think about what this would mean for people updating to 8 as they might have a cluster which was upgraded consecutively from 2.x ->5.x -> 6.x -> 7.x -> 8.x and where up to 7.x , the roles that had the `fields` field would be happily allowed and working and will now suddenly silently not apply. I'll get back to you with ideas next week once I have a few more hours to think this through

LMK if you have more on this. I believe my current revision will log something like 'ElasticsearchParseException: failed to parse indices privileges for role [test]. unexpected field [fields]' onto the elasticsearch host - we could instead log a more helpful message like e.g. "rejected invalid role; the 2.x field permissions syntax (\"fields\": [...]) has been deprecated", maybe even including a pointer to https://www.elastic.co/guide/en/elasticsearch/reference/current/field-level-security.html or some other migration guide. Or, probably overkill - is there precedent for returning detailed error messages to the client?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java [hunk 5]
```diff
@@ -25,11 +25,11 @@ public void testBWCFieldPermissions() throws Exception {
         Path path = getDataPath("roles2xformat.json");
         byte[] bytes = Files.readAllBytes(path);
         String roleString = new String(bytes, Charset.defaultCharset());
-        try (Client client = new NoOpClient("testBWCFieldPermissions")) {
+        try (Client client = new NoOpClient("test2xFieldPermissions")) {
```
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java:28]: This actually tests that your changes here are applied, i.e. that we reject the `fields` field so let's keep this for now.

[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java:28]: Note that you need to keep `roles2xformat.json` too and not delete it

**Comment by jkakavas** [comment]
> Should I force push a correction? 

Not sure how CLA checker will handle this but you can try. Worse case scenario, we'll close this PR and you would need to open a new one :/ 

> LMK if you have more on this. I believe my current revision will log something like 'ElasticsearchParseException: failed to parse indices privileges for role [test]. unexpected field [fields]' onto the elasticsearch host - we could instead log a more helpful message like e.g. "rejected invalid role; the 2.x field permissions syntax ("fields": [...]) has been deprecated", maybe even including a pointer to https://www.elastic.co/guide/en/elasticsearch/reference/current/field-level-security.html or some other migration guide. Or, probably overkill - is there precedent for returning detailed error messages to the client?

We are having an internal discussion about how this should be handled in general. Depending on the outcome, we might need to stall merging this until we have an acceptable solution to protect users from unexpected behaviors when upgrading to 8. I will keep you posted

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjaminran) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/role/PutRoleBuilderTests.java:28]: I deleted xpack/security/authz/store/roles2xformat.json (was used by NativeRolesStoreTests.testBWCFieldPermissions before I deleted that), but left xpack/security/action/role/roles2xformat.json in place so this test still passes - I think that's what you meant. LMK if not.

I don't see the difference between this test and the other tests I deleted like NativeRolesStoreTests.testBWCFieldPermissions. They all seem redundant after the test case I added in org.elasticsearch.xpack.security.authz.RoleDescriptorTests#testParse (`// we should reject a role where field permissions are stored in 2.x format`...). LMK if I'm missing something.

**Comment by benjaminran** [comment]
Okay, force pushed with corrected author names, rebased the 3 commits I've made on top of master, no code changes. Sorry about the extra noise

**Pull request closed without merging**